MINI-PROJECTS.md

```diff
@@ -104,6 +104,314 @@ share a bound.
+
+Most users of CBMC have access to multi-core machines; many have clusters available. There
+are a number of different approaches and a number of different levels at which the task
+could be parallelised; per assertion, portfolio approaches of different back-ends,
+parallel solvers, etc. It would be good to be able to exploit these. The caching approach
+suggested under “KLEE-PROVER” might be an easy route to implementing this.
+
+### x86 to Goto
+
+Look at various efforts to model x86 assembler, e.g., Andrew Kennedy,
+“Formalizing .EXEs, .DLLs, and all that”, Gang Tan, “Reusable tools for formal modeling of machine code”,
+and check whether this can turn into a translator from x86 to goto binary.
+
+Also look at Google's Native Client, and talk to Matt, who has tried Valgrind.
+
+Furthermore, talk to Michael, who has got a student who made some progress on this.
+
+And maybe have a read of Automated synthesis of symbolic instruction encodings from I/O samples.
+
+### Branch Prediction
+
+Implement branch prediction algorithms in path-symeex.
+
+### An abstract interpreter for C
+
+There's a lack for a robust, open source, abstract interpreter for C/C++. CPROVER contains most of the
+components for it, the key thing that would be needed is a front-end and some testing, plus possibly
+some more domains.
+
+### GCC Torture Tests
+
+GCC has a set of 'torture tests' which are intended to stress test and find bugs in their C parser
+and front-end. It would be good to run these through goto-cc / add them to the nightly testing and
+fix any bugs that arise. (Possible overlap with “Front-End Testing” above.). The same is probably
+also true for CSmith and other compiler fuzzing tools.
+
+### Compiler loop optimisations for BMC
+
+Try out some of the compiler loop optimisations (hoisting invariant code,
+fusing sequential independent loops, etc.) and see if they improve BMC,
+K-Induction or loop acceleration. These should be possible to implement in goto-instrument.
+
+Michael has an undergraduate student looking at this… (?)
+
+### Handling restrict in a smart way
+
+C99 adds the `restrict` keyword to mark unaliased pointers. This can be discarded by the compiler
+and thus has no semantics, it is only an annotation. However, there are a number of thing we could
+do with:
+
+1. Insert checks in calling functions to make sure that the pointers really are unaliased.
+2. Add an instrumentation pass that adds assumptions that restricted pointers are not aliased.
+This may imporve the quality of counter-examples (although at the cost of potentially missing
+some bugs, although not if combined with the previous technique).
+
+### C++ Exceptions
+
+C++ has annotations for what exceptions a function can throw. It would be good to be
+able to check / generate / minimise these and it should be possible to do this with
+a reasonably simple abstract domain. The CVC4 project would be interested in using this if it exists.
+
+Peter may have some thoughts on this. (?)
```

Comment:
An old prototype of something similar is here: https://github.com/peterschrammel/cbmc/commits/progres-exceptions